File: support-models/src/main/scala/com/gu/support/workers/states/CreateZuoraSubscriptionState.scala
package com.gu.support.workers.states

import com.gu.i18n.Country
import com.gu.salesforce.Salesforce.SalesforceContactRecords
import com.gu.support.acquisitions.AcquisitionData
import com.gu.support.encoding.Codec.deriveCodec
import com.gu.support.encoding.{Codec, DiscriminatedType}
import com.gu.support.workers.{PaymentMethod, SalesforceContactRecord, User, _}
import org.joda.time.LocalDate
import com.gu.support.encoding.CustomCodecs.{decodeCountry, decodeLocalTime, encodeCountryAsAlpha2, encodeLocalTime}
import com.gu.support.encoding.{Codec, DiscriminatedType}
import com.gu.support.workers._
import org.joda.time.LocalDate

import java.util.UUID

case class CreateZuoraSubscriptionState(
    productSpecificState: CreateZuoraSubscriptionProductState,
    requestId: UUID,
    user: User,
    product: ProductType,
    analyticsInfo: AnalyticsInfo,
    firstDeliveryDate: Option[LocalDate],
    appliedPromotion: Option[AppliedPromotion],
    csrUsername: Option[String],
    salesforceCaseId: Option[String],
    acquisitionData: Option[AcquisitionData],
) extends FailureHandlerState

object CreateZuoraSubscriptionState {
  implicit val codec: Codec[CreateZuoraSubscriptionState] = deriveCodec[CreateZuoraSubscriptionState]
}

sealed trait CreateZuoraSubscriptionProductState

object CreateZuoraSubscriptionProductState {

  case class ContributionState(
      product: Contribution,
      paymentMethod: PaymentMethod,
      salesForceContact: SalesforceContactRecord,
      similarProductsConsent: Option[Boolean],
  ) extends CreateZuoraSubscriptionProductState

  case class SupporterPlusState(
      billingCountry: Country,
      product: SupporterPlus,
      paymentMethod: PaymentMethod,
      appliedPromotion: Option[AppliedPromotion],
      salesForceContact: SalesforceContactRecord,
      similarProductsConsent: Option[Boolean],
  ) extends CreateZuoraSubscriptionProductState

  case class TierThreeState(
      user: User,
      product: TierThree,
      paymentMethod: PaymentMethod,
      firstDeliveryDate: LocalDate,
      appliedPromotion: Option[AppliedPromotion],
      salesForceContact: SalesforceContactRecord,
      similarProductsConsent: Option[Boolean],
  ) extends CreateZuoraSubscriptionProductState

  case class GuardianAdLiteState(
      product: GuardianAdLite,
      paymentMethod: PaymentMethod,
      salesForceContact: SalesforceContactRecord,
  ) extends CreateZuoraSubscriptionProductState

  case class DigitalSubscriptionState(
      billingCountry: Country,
      product: DigitalPack,
      paymentMethod: PaymentMethod,
      appliedPromotion: Option[AppliedPromotion],
      salesForceContact: SalesforceContactRecord,
      similarProductsConsent: Option[Boolean],
  ) extends CreateZuoraSubscriptionProductState

  case class PaperState(
      user: User,
      product: Paper,
      paymentMethod: PaymentMethod,
      firstDeliveryDate: LocalDate,
      appliedPromotion: Option[AppliedPromotion],
      salesForceContact: SalesforceContactRecord,
      similarProductsConsent: Option[Boolean],
  ) extends CreateZuoraSubscriptionProductState

  case class GuardianWeeklyState(
      user: User,
      giftRecipient: Option[GiftRecipient],
      product: GuardianWeekly,
      paymentMethod: PaymentMethod,
      firstDeliveryDate: LocalDate,
      appliedPromotion: Option[AppliedPromotion],
      salesforceContacts: SalesforceContactRecords,
      similarProductsConsent: Option[Boolean],
  ) extends CreateZuoraSubscriptionProductState

  import ExecutionTypeDiscriminators._

  private val discriminatedType = new DiscriminatedType[CreateZuoraSubscriptionProductState](fieldName)
  implicit val codec: Codec[CreateZuoraSubscriptionProductState] = discriminatedType.codec(
    List(
      discriminatedType.variant[ContributionState](contribution),
      discriminatedType.variant[SupporterPlusState](supporterPlus),
      discriminatedType.variant[DigitalSubscriptionState](digitalSubscription),
      discriminatedType.variant[PaperState](paper),
      discriminatedType.variant[GuardianWeeklyState](guardianWeekly),
      discriminatedType.variant[TierThreeState](tierThree),
      discriminatedType.variant[GuardianAdLiteState](guardianAdLite),
    ),
  )

}

File: support-workers/src/typescript/model/createZuoraSubscriptionState.ts
import { productPurchaseSchema } from '@modules/product-catalog/productPurchaseSchema';
import { appliedPromotionSchema } from '@modules/promotions/schema';
import { z } from 'zod';
import { salesforceContactRecordSchema } from '../services/salesforce';
import { countrySchema } from './address';
import { paymentMethodSchema } from './paymentMethod';
import {
	contributionProductSchema,
	digitalPackProductSchema,
	guardianAdLiteProductSchema,
	guardianWeeklyProductSchema,
	paperProductSchema,
	supporterPlusProductSchema,
	tierThreeProductSchema,
} from './productType';
import {
	baseStateSchema,
	giftRecipientSchema,
	userSchema,
} from './stateSchemas';

const baseProductStateSchema = z.object({
	paymentMethod: paymentMethodSchema,
	productInformation: productPurchaseSchema.nullish(),
	salesForceContact: salesforceContactRecordSchema,
});

export const contributionStateSchema = z
	.object({
		productType: z.literal('Contribution'),
		product: contributionProductSchema,
		similarProductsConsent: z.boolean().nullable(),
	})
	.merge(baseProductStateSchema);

export type ContributionState = z.infer<typeof contributionStateSchema>;

export const supporterPlusStateSchema = z
	.object({
		productType: z.literal('SupporterPlus'),
		billingCountry: countrySchema,
		product: supporterPlusProductSchema,
		appliedPromotion: appliedPromotionSchema.nullable(),
		similarProductsConsent: z.boolean().nullable(),
	})
	.merge(baseProductStateSchema);
export type SupporterPlusState = z.infer<typeof supporterPlusStateSchema>;

export const tierThreeStateSchema = z
	.object({
		productType: z.literal('TierThree'),
		user: userSchema,
		product: tierThreeProductSchema,
		firstDeliveryDate: z.string(),
		appliedPromotion: appliedPromotionSchema.nullable(),
		similarProductsConsent: z.boolean().nullable(),
	})
	.merge(baseProductStateSchema);
export type TierThreeState = z.infer<typeof tierThreeStateSchema>;

export const guardianAdLiteStateSchema = z
	.object({
		productType: z.literal('GuardianAdLite'),
		product: guardianAdLiteProductSchema,
	})
	.merge(baseProductStateSchema);
export type GuardianAdLiteState = z.infer<typeof guardianAdLiteStateSchema>;

export const digitalSubscriptionStateSchema = z
	.object({
		productType: z.literal('DigitalSubscription'),
		billingCountry: countrySchema,
		product: digitalPackProductSchema,
		appliedPromotion: appliedPromotionSchema.nullable(),
		similarProductsConsent: z.boolean().nullable(),
	})
	.merge(baseProductStateSchema);
export type DigitalSubscriptionState = z.infer<
	typeof digitalSubscriptionStateSchema
>;

export const paperStateSchema = z
	.object({
		productType: z.literal('Paper'),
		user: userSchema,
		product: paperProductSchema,
		firstDeliveryDate: z.string(),
		appliedPromotion: appliedPromotionSchema.nullable(),
		similarProductsConsent: z.boolean().nullable(),
	})
	.merge(baseProductStateSchema);
export type PaperState = z.infer<typeof paperStateSchema>;

export const guardianWeeklyStateSchema = z
	.object({
		productType: z.literal('GuardianWeekly'),
		user: userSchema,
		giftRecipient: giftRecipientSchema.nullable(),
		product: guardianWeeklyProductSchema,
		firstDeliveryDate: z.string(),
		appliedPromotion: appliedPromotionSchema.nullable(),
		similarProductsConsent: z.boolean().nullable(),
	})
	.merge(baseProductStateSchema);
export type GuardianWeeklyState = z.infer<typeof guardianWeeklyStateSchema>;

export const createZuoraSubscriptionProductStateSchema = z.discriminatedUnion(
	'productType',
	[
		contributionStateSchema,
		supporterPlusStateSchema,
		tierThreeStateSchema,
		guardianAdLiteStateSchema,
		digitalSubscriptionStateSchema,
		paperStateSchema,
		guardianWeeklyStateSchema,
	],
);

export type ProductSpecificState = z.infer<
	typeof createZuoraSubscriptionProductStateSchema
>;

export const createZuoraSubscriptionStateSchema = baseStateSchema.merge(
	z.object({
		productSpecificState: createZuoraSubscriptionProductStateSchema,
	}),
);

export type CreateZuoraSubscriptionState = z.infer<
	typeof createZuoraSubscriptionStateSchema
>;
